ktor-network/posix/src/io/ktor/network/sockets/DatagramSocketImpl.kt

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2014-2020 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.network.sockets
+
+import io.ktor.network.selector.*
+import io.ktor.network.util.*
+import io.ktor.util.network.*
+import io.ktor.utils.io.*
+import io.ktor.utils.io.core.*
+import io.ktor.utils.io.errors.*
+import kotlinx.cinterop.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.channels.*
+import platform.posix.*
+import kotlin.coroutines.*
+
+internal class DatagramSocketImpl(
+    descriptor: Int,
+    selector: SelectorManager,
+    private val _localAddress: NetworkAddress,
+    private val _remoteAddress: NetworkAddress?,
+    parent: CoroutineContext = EmptyCoroutineContext
+) : BoundDatagramSocket, ConnectedDatagramSocket, ReadWriteSocket,
+    BaseNativeSocket(descriptor, selector, parent) {
+
+    override val localAddress: NetworkAddress
+        get() = _localAddress
+    override val remoteAddress: NetworkAddress
+        get() = _remoteAddress!! // TODO: What should happen here?
+
+    private val sender = BroadcastChannel<Datagram>(Channel.BUFFERED)
+
+    private val receiver = produce<Datagram>(coroutineContext) {
+        while (true) {
+            val received = receiveImpl()
+            channel.send(received)
+        }
+    }
+
+    init {
+        makeShared()
+
+        launch {
+            sender.consumeEach { datagram ->
+                sendImpl(datagram)
+            }
+        }
+    }
+
+    private tailrec suspend fun sendImpl(
+        datagram: Datagram,
+        bytes: ByteArray = datagram.packet.readBytes()
+    ) {
+        var result: Int? = null
+        bytes.usePinned { pinned ->
+            datagram.address.address.nativeAddress { address, addressSize ->
+                result = sendto(
+                    descriptor,
+                    pinned.addressOf(0),
+                    bytes.size.convert(),
+                    0,
+                    address,
+                    addressSize
+                ).toInt()
+            }
+        }
+        result!!
```

Comment:
Is there a better way to prevent `!!`?

Reply:
I would check `sendto` result for error and throw an exception
Also please consider using `?: error(...)`  with the good description instead of `!!`

Reply:
> I would check sendto result for error and throw an exception

Already handled a few lines below.

> Also please consider using `?: error(...)` with the good description instead of `!!`

Done, but not sure what the error description should be. The variable can never be null. The only reason it is nullable is because `usePinned` and `nativeAddress` do not have contracts that their lambda is always executed once.